File: app/src/main/java/me/yxp/qfun/hook/api/OnSendMsg.java
package me.yxp.qfun.hook.api;

import java.lang.reflect.Method;
import java.util.ArrayList;

import me.yxp.qfun.hook.base.ApiHookItem;
import me.yxp.qfun.utils.hook.HookUtils;
import me.yxp.qfun.utils.reflect.ClassUtils;
import me.yxp.qfun.utils.reflect.MethodUtils;

public final class OnSendMsg extends ApiHookItem {
    public static final ApiHookItem INSTANCE = new OnSendMsg();

    @Override
    public void loadHook() throws Throwable {
        Method sendMsg = MethodUtils.create(ClassUtils._IKernelMsgService$CppProxy())
                .withMethodName("sendMsg")
                .findOne();

        HookUtils.hookAlways(sendMsg, param -> {
            ArrayList<?> msgElements = (ArrayList<?>) param.args[2];
            for (Listener listener : mListenerSet) {
                if (listener instanceof SendMsgListener) {
                    ((SendMsgListener) listener).onSend(msgElements);
                }
            }
        }, null);
    }

    public interface SendMsgListener extends Listener {
        void onSend(ArrayList<?> msgElements) throws Throwable;
    }
}

File: app/src/main/java/me/yxp/qfun/hook/api/OnSendMsg.kt
package me.yxp.qfun.hook.api

import com.tencent.qqnt.kernel.nativeinterface.IKernelMsgService
import com.tencent.qqnt.kernel.nativeinterface.MsgElement
import me.yxp.qfun.annotation.HookItemAnnotation
import me.yxp.qfun.hook.base.BaseApiHookItem
import me.yxp.qfun.hook.base.Listener
import me.yxp.qfun.utils.hook.hookBefore
import me.yxp.qfun.utils.reflect.findMethod

@HookItemAnnotation("监听发送消息")
object OnSendMsg : BaseApiHookItem<SendMsgListener>() {


    @Suppress("UNCHECKED_CAST")
    override fun loadHook() {
        IKernelMsgService.CppProxy::class.java
            .findMethod {
                name = "sendMsg"
            }.hookBefore(this) { param ->

                val elements = param.args[2] as ArrayList<MsgElement>
                forEachChecked { it.onSend(elements) }
            }
    }

}

fun interface SendMsgListener : Listener {
    fun onSend(elements: ArrayList<MsgElement>)
}